### PR TITLE
Predefined Optimization Levels & Web-based Visual Debugger

### PR DESCRIPTION
**Pull request by reiner-dolp**
- [x] remove fixpoint iteration. Plan is to integrate this with hennings rewrite that does not require alternating passes. 
- [x] add compiler levels `--optimization none|moderate|aggressive|custom` which define a fixed sequence of optimizations. A custom list of optimizations now looks as follows: `--optimization custom:ConstantFolding.vcg,ConstantFolding`
- [ ] Add a functionality to automatically skip compiler phases when its obvious that they will not result in further improvements, e.g. ConstantFolding after AlgebraicSymplification is useless if no replacement with `*0` was done. (Postponed until we have several optimizations that need this functionality.)
- [x] Add optimization flag `gui` that will allow you to step through an optimization using henning's GUI.
   - [x] stop anywhere using `breakpoint!()` macro, show compiler state in GUI
   - [x] make styling of individual nodes possible (e.g. style current or marked nodes)
   - [x] make label of each node editable 
   - [x] hide debugger behind a rust compiler flag
   - [ ] make breakpoints filterable (will be part of a follow up PR that improves UX after the tool has been used for some time)

### CODE REVIEW

File: compiler-cli/src/optimization_arg.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
+//! Parses the optimization command line flag, which might look as follows:
+//!
+//! `--optimization Aggressive`
+//!
+//! or in development builds:
+//!
+//! `-O Custom|ConstantFolding:vcg:debug,ConstantFolding`
```
Comment (reiner-dolp) [inline compiler-cli/src/optimization_arg.rs]: outdated example

File: compiler-cli/src/optimization_arg.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+#[derive(Debug, Fail)]
+pub enum ParseError {
+    #[fail(display = "unknown optimization level '{}'", name)]
+    UnknownLevel { name: String },
+    #[fail(display = "unknown optimization pass '{}'", name)]
+    UnknownOptimization { name: String },
+    #[fail(display = "unknown optimization flag ':{}'", name)]
+    UnknownFlag { name: String },
+    #[fail(display = "optimization flag can only contain a single pipe")]
```
Comment (reiner-dolp) [inline compiler-cli/src/optimization_arg.rs]: optimization flag -> argument

Comment (reiner-dolp) [reply compiler-cli/src/optimization_arg.rs]: | -> :

File: compiler-cli/src/optimization_arg.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    TooManySeparators,
+    #[fail(display = "a custom optimization sequence requires a list of optimizations")]
```
Comment (reiner-dolp) [inline compiler-cli/src/optimization_arg.rs]: a non-empty list of

File: compiler-lib/src/optimization/mod.rs [hunk 4]
```diff
@@ -17,89 +50,134 @@ pub use self::fixpoint::Fixpoint;
...
+pub enum Level {
+    /// This level compiles the fastest and generates the most debuggable code
+    /// since it does not perform any optimization.
+    None,
+    /// Moderate level of optimization which enables most optimizations.
+    Moderate,
+    /// Enables optimizations that take longer to perform or that may generate
+    /// larger code in an attempt to make the program run faster.
+    Aggressive,
+    /// Enables optimizations that take longer to perform or that may generate
```
Comment (reiner-dolp) [inline compiler-lib/src/optimization/mod.rs]: wrong comment

File: compiler-lib/src/optimization/mod.rs [hunk 5]
```diff
@@ -17,89 +50,134 @@ pub use self::fixpoint::Fixpoint;
...
+}
+
+impl Default for Level {
+    fn default() -> Level {
+        Level::None
+    }
+}
+
+impl Level {
+    fn sequence(&self) -> Vec<Optimization> {
```
Comment (reiner-dolp) [inline compiler-lib/src/optimization/mod.rs]: add doc comment

File: compiler-lib/src/asciifile/line_number_cache.rs [hunk 6]
```diff
@@ -92,10 +92,10 @@ impl LineNumberCache {
     /// are out of range. Prefer the safe API of the `Position` struct
     /// instead.
     pub fn row(&self, byte_offset: usize) -> usize {
-        // - binary_search always gives the index of the smaller number
-        //   if the fiven offset lies between two newlines.
-        // - binary_search returns len() if the number is bigger than
-        //   everything seen
+        //
+        // - binary_search always gives the index of the smaller number if the fiven
```
Comment (flip1995) [inline compiler-lib/src/asciifile/line_number_cache.rs]:
```suggestion
         // - binary_search always gives the index of the smaller number if the given
```

File: compiler-cli/Cargo.toml [hunk 7]
```diff
@@ -35,3 +35,4 @@ env_logger = "0.6"
 # Enable all UB-safe-making features by default.
 default = [ "checked_new" ]
 checked_new = [ "mjrt/checked_new" ]
+gui_debugger = [ "compiler-lib/gui_debugger" ]
```
Comment (flip1995) [inline compiler-cli/Cargo.toml]: Isn't this called `debugger_gui`?

File: compiler-lib/Cargo.toml [hunk 8]
```diff
@@ -33,3 +38,10 @@ criterion = "0.2"
 [[bench]]
 name = "symtab_bench"
 harness = false
+
+[features]
+default = [ ]
+# This feature enables the `breakpoint!` macro, which
+# among other things, is used to implement the `gui`
+# flag in custom optimization sequences.
+gui_debugger = [ ]
```
Comment (flip1995) [inline compiler-lib/Cargo.toml]: Also `gui_debugger` vs `debugger_gui`. I would prefer a consistent naming.

File: compiler-lib/src/debugging.rs [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+        );
+    }};
+}
+
+#[cfg(not(feature = "gui_debugger"))]
+#[macro_export]
+macro_rules! breakpoint {
+    ($label:expr, $prog:expr) => {{
+        // TODO: there has to be a better way
+        fn use_macro_argument<T>(_val: &T) {}
```
Comment (flip1995) [inline compiler-lib/src/debugging.rs]: `let _ = $val`

**Comment by reiner-dolp** [comment]
bors r=flip1995

**Pull request merged**